File: Longest-ZigZag-Path-in-a-Binary-Tree.py
 # Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        
        output = 0
        fringe = []
        fringe.append((root,0,""))
        
        while fringe:
            
            node, index, path = fringe.pop()
            
            if node.left:
                
                cur = 1
                
                if path != "l":
                    
                    cur = index+1
                
                fringe.append((node.left, cur, "l"))         
                output = max(output, cur)
            
            if node.right:
                
                cur = 1
                                
                if path != "r":
                    
                    cur = index+1
                
                fringe.append((node.right, cur, "r"))
                output = max(output,cur)
                
        return output
        
        
        

File: Longest-ZigZag-Path-in-a-Binary-Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    int output = 0;
    
    private void rec(TreeNode node, int count, char path){
        
        if (node == null){
            
            return;
            
        }
        
        output = Math.max(output,count);
        
        if (path == 'r') {
            
            rec(node.left,count+1,'l');
            rec(node.right,1,'r');
            
        }
        
        else if (path == 'l') {
            
            rec(node.left,1,'l');
            rec(node.right,count+1,'r');
            
        }
        
        else {
            
            rec(node.left,1,'l');
            rec(node.right,1,'r');
            
        }
        
    }
    
    public int longestZigZag(TreeNode root) {
        
        rec(root,0,'c');
        
        return output;
        
    }
}
